### ISSUE TITLE
SPEI Predictions: how is it possible that there is a plot for 100% for city_for_training?

### ISSUE THREAD

**Issue by A-Infor** [A-Infor/Python-OOP-LSTM-Drought-Predictor#48]
1. The data is split into 80% and 20%;
2. The model trains on 80% and predicts 20% (test);
3. So, how is it possible that there is a plot showing predictions for 100% of the data?

<img width="432" height="288" alt="Image" src="https://github.com/user-attachments/assets/035889bf-87cb-4602-8cac-01196855eb4e" />

Wouldn't the correct be to have only 20% predictions?

<img width="432" height="288" alt="Image" src="https://github.com/user-attachments/assets/f8d53eb9-942b-4673-b618-43cc151ed98a" />

**Comment by A-Infor** [comment]
In the Drought repository, `showPredictionResults()` has these lines:
```
trueValues  = np.append(trainDataTrueValues, testDataTrueValues)
predictions = np.append(trainPredictValues    , testPredictValues   )
```
It is pretty clear that there are values predicted during training, not only during testing. These come from `UseNeuralNetwork()`:
```
trainPredictValues = model.predict(trainDataForPrediction)
testPredictValues  = model.predict(testDataForPrediction )
```
But, why there is this prediction on the 80% data it already trained on?

**Comment by A-Infor** [comment]
> But, why there is this prediction on the 80% data it already trained on?

Both ChatGPT and Gemini think it's OK to predict values of the same dataset used for training:

1. it helps visualize if the model learned the data patterns;
2. makes it possible to compare testing and training errors to detect overfitting;
3. to help visualize predictions x reality in the training dataset, but not as a final evaluation.

It's important to notice that `model.fit()` reports loss values during training, but only `model.predict()` is able to tell the FINAL performance of the model after it is done training. So `model.fit()` is for optimization, while `model.predict()` is for evaluation.

Thus, "it's OK to predict values of the same dataset used for training". It is not a logical/conceptual error. But it serves for some purposes, and not for others --it is not a trustworthy way to evaluate the model generalization capacity. It just shows if the training data was learnt. It helps see the model performance, but could lead to false conclusions, necessitating the testing metrics.

**Comment by A-Infor** [comment]
I've read the articles recommended by Gemini on why it's OK to predict values for the same dataset:

1. https://www.comet.com/site/blog/why-is-model-evaluation-important-in-machine-learning/
2. https://medium.com/@mookpreeyanuch/a-begunners-guide-to-model-evaluation-in-machine-learning-6e4ad4a844b2
3. https://www.kaggle.com/discussions/getting-started/466697
4. https://medium.com/@petraclara20/overfitting-in-machine-learning-how-to-detect-and-prevent-it-5ae4ed2ec7e4
5. https://en.wikipedia.org/wiki/Training,_validation,_and_test_data_sets

They helped me nothing to understand deeper why this repository's code is not wrong.

**Comment by A-Infor** [comment]
I've read the articles recommended by ChatGPT and they are also completely useless. What a waste of time!

1. https://www.researchgate.net/post/Why_should_the_accuracies_of_both_train_and_test_data_sets_be_included_in_the_evaluation_metric_of_a_neural_network
2. https://www.reddit.com/r/learnmachinelearning/comments/1htlaiv/how_does_a_trained_ml_model_make_real_world/
3. https://www.wired.com/story/ai-pioneer-algorithms-understand-why/
4. https://stats.stackexchange.com/questions/278800/predict-using-trained-model-on-dataset
5. https://datascience.stackexchange.com/questions/94342/whats-the-point-with-neural-networks-if-you-can-only-predict-linear-test-data
6. https://en.wikipedia.org/wiki/Generalization_error
7. https://en.wikipedia.org/wiki/Overfitting
8. https://en.wikipedia.org/wiki/Training%252C_validation%252C_and_test_data_sets
9. https://arxiv.org/pdf/q-bio/0510017
10. https://datascience.stackexchange.com/questions/86632/why-is-it-wrong-to-train-and-test-a-model-on-the-same-dataset
11. https://stats.stackexchange.com/questions/278800/predict-using-trained-model-on-dataset

**Comment by A-Infor** [comment]
ChatGPT pointed out, and Gemini confirmed, that these terms are relevant for this issue:

- Resubstituion Error

- Training Error

- Training Set Error

- In-Sample Error

- Apparent Error

- Empirical Error

- Empirical Risk 

I'm already searching for articles on Medium.com about these terms.

**Comment by A-Infor** [comment]
> * Resubstituion Error

No useful results for "Resubstitution"/"Re-substituion"/"Re substitution" on Medium.

**Comment by A-Infor** [comment]
I've read these 3 books currently available on https://minhabiblioteca.com.br/:

1. Python para Data Science e Machine Learning Descomplicado -- Amilcar Netto; Francisco Maciel -- Editora Alta Books;
2. Projetando sistemas de Machine Learning: processo interativo para aplicações prontas para produção -- Chip Huyen -- Editora Alta Books;
3. Técnicas de machine learning -- Abraham Laredo Sicsú; André Samartini; Nelson Lerner Barth -- Editora Blucher.

Only the third said something useful about training x testing portions of datasets. It mentioned that it is not a good practice to take in consideration the training portion performance as the real performance of the model.

These 2 books currenytly available on https://bvirtual.com.br were useless to me:

1. Machine Learning: Introdução à Classificação;
2. Introdução à Data Science: Algoritmos de Machine Learning e Métodos de Análise.

**Comment by A-Infor** [comment]
I've read this only book available on https://bvirtual.com.br/ and it was useless:

1. Inteligência artificial e aprendizagem de máquina: aspectos teóricos e aplicações.

I also read these other books available on https://minhabiblioteca.com.br/, and they were equally useless:

1. Blueprints de aprendizado de máquina e ciência de dados para finanças: desenvolvendo desde estratégias de trades até robôs Advisors com Python;
2. Inteligência Artificial - Uma Abordagem de Aprendizado de Máquina (3ª edição);
3. Inteligência Artificial - Uma Abordagem de Aprendizado de Máquina (2ª edição).

**Comment by A-Infor** [comment]
I've read these books from https://minhabiblioteca.com.br/, and they were useless:
1. A Revolução do Aprendizado Profundo;
2. Aprendizado profundo para leigos.

There were no corresponding books on https://bvirtual.com.br/.

**Comment by A-Infor** [comment]
I've checked all these articles, searching for explanations about using or not using the training dataset for making predictions:

1. Short-term drought Index forecasting for hot and semi-humid climate Regions: A novel empirical Fourier decomposition-based ensemble Deep-Random vector functional link strategy
2. An election algorithm combined with support vector regression for estimating hydrological drought
3. Integration of extreme learning machines with CEEMDAN and VMD techniques in the prediction of the multiscalar standardized runoff index and standardized precipitation evapotranspiration index
4. High Resolution Forecasting of Summer Drought in the Western United States
5. Standardized precipitation evapotranspiration index (SPEI) estimated using variant long short-term memory network at four climatic zones of China
6. Comparison of Machine Learning Models for Drought Prediction in Central Myanmar
7. Spatio-temporal distribution and prediction of agricultural and meteorological drought in a Mediterranean coastal watershed via GIS and machine learning
8. Rainfall Forecast and Drought Analysis for Recent and Forthcoming Years in India
9. Combining deep learning methods and multi-resolution analysis for drought forecasting modeling
10. Comparison of hybrid machine learning models to predict short-term meteorological drought in Guanzhong region, China
11. A Novel Approach for High-Performance Estimation of SPI Data in Drought Prediction
12. Improving drought modeling based on new heuristic machine learning methods
13. Advanced Machine Learning Model for Prediction of Drought Indices using Hybrid SVR-RSM
14. A comparative analysis of data mining techniques for agricultural and hydrological drought prediction in the eastern Mediterranean
15. Ensemble Learning with CNN and BMO for Drought Prediction
16. Drought monitoring and prediction using SPI, SPEI, and random forest model in various climates of Iran
17. Drought Forecasting in Alibori Department in Benin using the Standardized Precipitation Index and Machine Learning Approaches
18. Drought forecasting using new advanced ensemble-based models of reduced error pruning tree
19. Deep Learning Oriented Satellite Remote Sensing for Drought and Prediction in Agriculture
20. A machine learning model for drought tracking and forecasting using remote precipitation data and a standardized precipitation index from arid regions
21. An improved SPEI drought forecasting approach using the long short-term memory neural network
22. A novel hybrid dragonfly optimization algorithm for agricultural drought prediction
23. Interpretable and explainable AI (XAI) model for spatial drought prediction
24. Fusion-based framework for meteorological drought modeling using remotely sensed datasets under climate change scenarios: Resilience, vulnerability, and frequency analysis
25. Forecasting standardized precipitation index using data intelligence models: regional investigation of Bangladesh
26. Zoning map for drought prediction using integrated machine learning models with a nomadic people optimization algorithm
27. Comparison of wavelet and empirical mode decomposition hybrid models in drought prediction
28. Long-Lead Predictions of Warm Season Droughts in South Korea Using North Atlantic SST
29. Forecasting Drought Indices Using Machine Learning Algorithm
30. Meteorological drought forecasting based on a statistical model with machine learning techniques in Shaanxi province, China
31. Evaluation of performance of drought prediction in Indonesia based on TRMM and MERRA-2 using machine learning methods
32. Drought Prediction for Areas with Sparse Monitoring Networks: A Case Study for Fiji
33. Input selection and data-driven model performance optimization to predict the Standardized Precipitation and Evaporation Index in a drought-prone region
34. Meteorological drought forecasting for ungauged areas based on machine learning: Using long-range climate forecast and remote sensing data
35. Application of the extreme learning machine algorithm for the prediction of monthly Effective Drought Index in eastern Australia

I've found this useful, but short explanation, on the article "Forecasting drought indices using machine learning algorithm":

> The RMSE from the training dataset indicates how close the model fits the training dataset, while the one from testing shows how good the model predicts. High training RMSE is an indication of underfitting, while low training RMSE with high testing RMSE is happening for overfitting. Therefore, the lower in both training and testing RMSE, the better is fitting model

I was also able to get nice formulations of RMSE, MSE, MAE and R² on "Drought Forecasting in Alibori Department in Benin using the Standardized Precipitation Index and Machine Learning Approaches".

When reading these 35 articles, I could not find information about some ML models being better at predicting intense droughts, and other ML models being better at predicting mild droughts. I did read again the summaries I previously wrote about these same 35 articles and could not find information either. I think I did read this not abor drought **prediction**, but about drought **classification**, and that, thus, was discarded as not relevant. So, I will not search for this kind of information anymore.

But I still want a better explanation on what to do with performance metrics of predictions made over the training dataset.

**Comment by A-Infor** [comment]
I've checked the paid catalog of these 2 e-learning websites and they are not useful to understand beyond the coding of machine learning models:

1. https://www.codecademy.com/
2. https://web.dio.me/

**Comment by A-Infor** [comment]
I've found this useful explanation, on the book "Deep learning with Python: develop deep learning models on Theano and TensorFlow using Keras":

> `We have trained our neural network on the entire dataset and we can evaluate the performance of the network on the same dataset. This will only give us an idea of how well we have modeled the dataset (e.g. train accuracy), but no idea of how well the algorithm might perform on new data. We have done this for simplicity, but ideally, you could separate your data into train and test datasets for the training and evaluation of your model.''
> 
> ``You can evaluate your model on your training dataset using the evaluation() function on your model and pass it the same input and output used to train the model. This will generate a prediction for each input and output pair and collect scores, including the average loss and any metrics you have configured, such as accuracy.''

**Comment by A-Infor** [comment]
> why there is this prediction on the 80% data it already trained on?

What is clear so far?

1. there are metrics for training and for testing because it's normal to make predictions based on both portions of the dataset;
2. the predictions based on the training data is just to check if the model learnt well enough, and for comparing with the testing metrics;
3. the testing metrics are the ones that are really relevant.

I think that, with these informatons, this question is already solved.
